cqrs-codegen/src/event/versioned_event.rs

```diff
@@ -0,0 +1,168 @@
+//! Codegen for [`cqrs::VersionedEvent`]
+
+use quote::quote;
+use syn::{
+    parse::{Error, Result},
+    punctuated::Punctuated,
+    spanned::Spanned as _,
+};
+use synstructure::Structure;
+
+use crate::{event, util};
+
+/// Implements [`crate::derive_versioned_event`] macro expansion.
+pub fn derive(input: syn::DeriveInput) -> Result<proc_macro2::TokenStream> {
+    match input.data {
+        syn::Data::Struct(_) => derive_struct(input),
+        syn::Data::Enum(_) => derive_enum(input),
+        syn::Data::Union(data) => Err(Error::new(
+            data.union_token.span(),
+            "Unions are not supported for deriving VersionedEvent",
+        )),
+    }
+}
+
+/// Implements [`crate::derive_versioned_event`] macro expansion for structs.
+pub fn derive_struct(input: syn::DeriveInput) -> Result<proc_macro2::TokenStream> {
+    let syn::DeriveInput {
+        attrs,
+        ident,
+        generics,
+        ..
+    } = input;
+
+    let meta = util::get_nested_meta(&attrs, "event")?.ok_or_else(|| {
+        Error::new(
+            ident.span(),
+            "Expected struct to have #[event(...)] attribute",
+        )
+    })?;
+
+    let const_val = parse_event_version_from_nested_meta(&meta)?;
+    let const_doc = format!("Version of [`{}`] event", ident);
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    Ok(quote! {
+        #[automatically_derived]
+        impl#impl_generics #ident#ty_generics #where_clause {
+            #[doc = #const_doc]
+            #[allow(unsafe_code)]
+            pub const EVENT_VERSION: ::cqrs::EventVersion =
+                unsafe { ::cqrs::EventVersion::new_unchecked(#const_val) };
```

Comment:
`cqrs::EventVersion::new_unchecked` is under "unsafe" feature flag. I don't know what is the best way to handle feature flags in generated code. Should we generate code that can handle conditional compilation, should we generate different implementations, depending on this flag or should we just leave it as-is and document, that derive-VersionedEvent requires `cqrs` "unsafe" feature enabled?

Reply:
@ffuugoo no, there is no need for conditional compilation here. When generating code we can check the actual value for being >0 and reject with error if it's not. After we've checked that generating such `unsafe` code will be safe.

However, we should consider to use safe version `cqrs::EventVersion::new` here with an appropriate unpacking. I'm unsure, however, whether current Rust const evaluation will allow that. If not, just remove the `unsafe` feature flag.